### DISCUSSION TITLE
Hydrated pending mutations are not continued

### DISCUSSION THREAD

**Discussion by momme-rtf** [TanStack/query#7044] [General]
We're working on a react native app in combination with react query that follows the offline first approach. In our app we have the use case of a long running mutation. If a user starts this mutation and kills the app from the background before the mutation was successful we would still want to resume this mutation as soon as the user reopens the app.
In the point in time when the user kills the app the mutation is in the following state:
- status: pending
- isPaused: false

Since the `defaultShouldDehydrateMutation` function makes sure that only paused mutations are being persisted our mutation would get ignored.
In order to persist it anyways we customized the `shouldDehydrateMutation` of the PersistQueryClientProvider:
`(m: Mutation) => defaultShouldDehydrateMutation(m) || m.state.status === 'pending'`
When reopening the app, the mutation will get hydrated but it won't continue it's mutationFn.
Do you think this should be handled by react query internally or do you see this case as too specific so that we'd need to call continue on them again manually in the onSuccess handler of the PersistQueryClientProvider?
```
<PersistQueryClientProvider
    client={queryClient}
    onSuccess={() => {
        // The following 4 lines solve the problem we have
        const mutations = queryClient.getMutationCache().getAll().filter((m) => !m.state.isPaused && (m.state.status === 'idle' || m.state.status === 'pending'));
        for (const mutation of mutations) {
            void mutation.continue();
        }

        if (onlineManager.isOnline()) {
            queryClient.resumePausedMutations()
        }
    }}
    persistOptions={{
        persister,
        dehydrateOptions: {
            shouldDehydrateMutation: (m: Mutation) => defaultShouldDehydrateMutation(m) || m.state.status === 'idle' || m.state.status === 'pending'
        }
    }}
>
    {props.children}
</PersistQueryClientProvider>
```

**Comment by TkDodo** [comment]
oh that's interesting. yeah we don't offer something to continue non-paused mutations out of the box, but your code should work fine.

**Reply by momme-rtf** [reply]
But would agree that non-paused mutations should also be resumed by react query internally?

**Reply by TkDodo** [reply]
no, not really. the default behaviour is to persist paused mutations and then we offer a function to resume paused mutations. If you want something else, you'd need to do it on your own, and we provide the primitives for that.

**Reply by ramonxm** [reply]
any solutions for this? @momme-rtf  @TkDodo I need to do the same thing

**Reply by momme-rtf** [reply]
@ramonxm our solution is already shown in the code snippet. Check out the lines below the comment.

**Reply by ramonxm** [reply]
> @ramonxmnossa solução já está mostrada no trecho de código. Confira as linhas abaixo do comentário.

I did this but I am getting this error: No mutationFn found

**Reply by ramonxm** [reply]
Ah, I need to use setMutationDefaults, with this it worked.
